jib-core/src/main/java/com/google/cloud/tools/jib/event/DefaultEventEmitter.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.event;
+
+import com.google.common.collect.ImmutableMultimap;
+
+/** Emits {@link JibEvent}s to event handlers. */
+public class DefaultEventEmitter implements EventEmitter {
+
+  /** Maps from {@link JibEvent} class to handlers for that event type. */
+  private final ImmutableMultimap<Class<? extends JibEvent>, Handler<? extends JibEvent>> handlers;
+
+  /**
+   * Creates an instance from {@link Handler}s in an {@link EventHandlers}.
+   *
+   * @param eventHandlers the {@link EventHandlers} to get the {@link Handler}s from
+   */
+  public DefaultEventEmitter(EventHandlers eventHandlers) {
+    handlers = eventHandlers.getHandlers();
+  }
+
+  /**
+   * Emits {@code jibEvent} to all the handlers that can handle it.
+   *
+   * @param jibEvent the {@link JibEvent} to emit
+   */
+  @Override
+  public void emit(JibEvent jibEvent) {
```

Comment:
Should this happen on an "event" thread? So slow event handlers don't slow down the build steps?

Could events should be submitted to the dispatcher and then it just watches its queue to dispatch them to handlers (on a single event handler thread?)

Reply:
I still don't get the exact workings of our eventing design, but from my experience on some event models, I think it is a sane design that the event handlers run on the same thread where the actual event is taking place and often it is desirable. That is, if an event handler blocks, the main execution path will block. Often, you want to make sure certain things happen in a synchronized manner with the event, and if the handlers run on a separate thread that is independent of the main execution thread, I think it will be almost impossible to sync. Think about UI a text input widget as example: you want to validate if what the user entered in a field is correct, and the validation is triggered through text input events.

Reply:
Yea, I think we should keep it on the same thread and delegate to the user to handle with inexpensive handlers. That would avoid the cost of dispatching to another thread for our plugins' case of just logging a log message.